outputs/aws.go

```diff
@@ -7,49 +7,69 @@ import (
 	"errors"
 	"fmt"
 	"log"
+	"net/http"
 	"net/url"
 	"os"
 	"time"
 
 	"github.com/DataDog/datadog-go/statsd"
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
 	"github.com/aws/aws-sdk-go/aws/session"
 	"github.com/aws/aws-sdk-go/service/cloudwatchlogs"
 	"github.com/aws/aws-sdk-go/service/lambda"
 	"github.com/aws/aws-sdk-go/service/s3"
 	"github.com/aws/aws-sdk-go/service/sns"
 	"github.com/aws/aws-sdk-go/service/sqs"
-	"github.com/aws/aws-sdk-go/service/sts"
 
 	"github.com/falcosecurity/falcosidekick/types"
 )
 
 // NewAWSClient returns a new output.Client for accessing the AWS API.
 func NewAWSClient(config *types.Configuration, stats *types.Statistics, promStats *types.PromStatistics, statsdClient, dogstatsdClient *statsd.Client) (*Client, error) {
 
-	if config.AWS.AccessKeyID != "" && config.AWS.SecretAccessKey != "" && config.AWS.Region != "" {
+	var region string
+	var err error
+
+	//Set the region, or lookup if it doesn't exist
+	if config.AWS.Region != "" {
+		region = config.AWS.Region
+	} else {
+		log.Printf("[INFO] : AWS - No Region configured, querying instance metadata for region \n")
+		region, err = GetRegion("")
+		if err != nil {
+			return nil, errors.New("error finding region from metadata, or configuration")
+		}
+	}
+	log.Printf("[INFO] : AWS - Using Region %v\n", region)
+
+	// If using access keys in the config, check for those first.
+	if config.AWS.AccessKeyID != "" && config.AWS.SecretAccessKey != "" {
 		err1 := os.Setenv("AWS_ACCESS_KEY_ID", config.AWS.AccessKeyID)
 		err2 := os.Setenv("AWS_SECRET_ACCESS_KEY", config.AWS.SecretAccessKey)
-		err3 := os.Setenv("AWS_DEFAULT_REGION", config.AWS.Region)
-		if err1 != nil || err2 != nil || err3 != nil {
+		if err1 != nil || err2 != nil {
 			log.Printf("[ERROR] : AWS - Error setting AWS env vars")
 			return nil, errors.New("Error setting AWS env vars")
 		}
+		// Log which type of authentication will be using.
+		log.Printf("[INFO] : AWS - Using Access Key Authentication")
 	}
 
-	sess, err := session.NewSession(&aws.Config{
-		Region: aws.String(config.AWS.Region)},
-	)
-	if err != nil {
-		log.Printf("[ERROR] : AWS - %v\n", "Error while creating AWS Session")
-		return nil, errors.New("Error while creating AWS Session")
+	//Check if using IRSA (IAM role service accounts)
+	// env | grep AWS
+	roleARN := os.Getenv("AWS_ROLE_ARN")
+	tokenPath := os.Getenv("AWS_WEB_IDENTITY_TOKEN_FILE")
```

Comment:
are these values mandatory? I thought it was possible to retrieve the credentials directly from metadata

Reply:
IRSA will put both of these as environment variables into the container from the eks-identity-webhook https://github.com/aws/amazon-eks-pod-identity-webhook

You could manually input these into the container, but the goal is for anyone using the amazon-eks-pod-identity-webhook can default use IRSA instead of node role for this.

Reply:
The default auth flow would be

1. Access Keys if added (Env vars exist)
2. IRSA if enabled (Env vars exist)
3. Attempt to create default session if neither exist... (Defaults to node role).

Reply:
ok, that's smart and cover all cases:
- instance profile
- service account mapped with an iam role
- keys

Reply:
Awesome!